[Module interface] Drop marker protocols from inheritance clauses for backward compatibility

DougGregor opened the pull request:
When printing a protocol definition whose inheritance clause includes
a marker protocol in a module interface, print an alternative definition
of the protocol that drops the marker protocol for compilers that
don't understand marker protocols. This only works for marker protocols
because they have no impact on the ABI. The end result is something like:

    #if compiler(>=5.3) && $MarkerProtocol
    public protocol MP3 : FeatureTest.MP {
    }
    #else
    public protocol MP3 {
    }
    #endif

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves SR-NNNN.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test macOS

slavapestov reviewed (commented):
It might be simpler if you printed marker protocols as a typealias to Any instead:

```
#if compiler(>=5.3) && $MarkerProtocol
@_marker protocol M {}
#else
typealias M = Any
#endif

protocol N : M {} // inheriting from M is a no-op on older compilers
```

slavapestov commented on test/ModuleInterface/features.swift:
```diff
@@ -42,9 +42,12 @@ public func globalAsync() async { }
 @_marker public protocol MP2: MP { }
 
 // CHECK: #if compiler(>=5.3) && $MarkerProtocol
```
Shouldn't this be compiler(>=5.4)?

DougGregor replied:
It doesn't really matter what version we use here; the point is that the `compiler` clause ensures that what's inside the `#if` region doesn't get parsed.

DougGregor commented:
> It might be simpler if you printed marker protocols as a typealias to Any instead:
> 
> ```
> #if compiler(>=5.3) && $MarkerProtocol
> @_marker protocol M {}
> #else
> typealias M = Any
> #endif
> 
> protocol N : M {} // inheriting from M is a no-op on older compilers
> ```

This is a fantastic idea! And way, way simpler than what I've been doing. Thanks!

DougGregor committed:
Simplify printing of marker protocols in Swift interfaces

Thanks to a great idea from Slava, simplify the printing of marker
protocols in Swift interfaces while maintaining backward compatibility.
For compilers that cannot handle marker protocols, print a typealias
to `Any` instead, which is a good stand-in for just about everything.

DougGregor commented:
Dropping this in favor of the simpler solution, https://github.com/apple/swift/pull/36049
